pkg/agent/controller/packetcapture/packetcapture_controller.go

```diff
@@ -0,0 +1,658 @@
+	}
+
+	switch pc.Status.Phase {
+	case "":
+		err = c.initPacketCapture(pc)
+	case crdv1alpha1.PacketCaptureRunning:
+		err = c.checkPacketCaptureStatus(pc)
+	default:
+		c.cleanupPacketCapture(pcName)
+	}
+	return err
+
+}
+
+func (c *Controller) getUploaderByProtocol(protocol StorageProtocolType) (ftp.Uploader, error) {
+	if protocol == sftpProtocol {
+		return c.sftpUploader, nil
+	}
+	return nil, fmt.Errorf("unsupported protocol %s", protocol)
+}
+
+func (c *Controller) generatePacketsPathForServer(name string) string {
+	return name + ".pcapng"
+}
+
+func getDefaultFileServerAuth() *crdv1alpha1.BundleServerAuthConfiguration {
+	return &crdv1alpha1.BundleServerAuthConfiguration{
+		AuthType: crdv1alpha1.BasicAuthentication,
+		AuthSecret: &v1.SecretReference{
+			Name:      fileServerAuthSecretName,
+			Namespace: fileServerAuthSecretNamespace,
+		},
+	}
+}
+
+func (c *Controller) uploadPackets(pc *crdv1alpha1.PacketCapture, outputFile afero.File) error {
+	if pc.Spec.FileServer == nil {
+		klog.V(2).Info("No fileserver info found in PacketCapture, skip upload packets file")
+		return nil
+	}
+	klog.V(2).InfoS("Uploading captured packets for PacketCapture", "name", pc.Name)
+	uploader, err := c.getUploaderByProtocol(sftpProtocol)
+	if err != nil {
+		return fmt.Errorf("failed to upload support bundle while getting uploader: %v", err)
+	}
+	authConfig := getDefaultFileServerAuth()
+	serverAuth, err := ftp.ParseBundleAuth(*authConfig, c.kubeClient)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get authentication for the fileServer", "name", pc.Name, "authentication", authConfig)
+		return err
+	}
+	cfg := ftp.GenSSHClientConfig(serverAuth.BasicAuthentication.Username, serverAuth.BasicAuthentication.Password)
+	return uploader.Upload(pc.Spec.FileServer.URL, c.generatePacketsPathForServer(pc.Name), cfg, outputFile)
+}
+
+// initPacketCapture mark the PacketCapture as running
+func (c *Controller) initPacketCapture(pc *crdv1alpha1.PacketCapture) error {
+	err := c.updatePacketCaptureStatus(pc, crdv1alpha1.PacketCaptureRunning, "", 0)
+	if err != nil {
+		return err
+	}
+	return c.startPacketCapture(pc)
+}
+
+func (c *Controller) updatePacketCaptureStatus(pc *crdv1alpha1.PacketCapture, phase crdv1alpha1.PacketCapturePhase, reason string, numCapturedPackets int32) error {
+	latestPC, err := c.packetCaptureLister.Get(pc.Name)
+	if err != nil {
+		return fmt.Errorf("get PacketCapture failed: %w", err)
+	}
+	type PacketCapture struct {
+		Status crdv1alpha1.PacketCaptureStatus `json:"status,omitempty"`
+	}
+	patchData := PacketCapture{Status: crdv1alpha1.PacketCaptureStatus{Phase: phase}}
+	if phase == crdv1alpha1.PacketCaptureRunning && pc.Status.StartTime == nil {
+		t := metav1.Now()
+		patchData.Status.StartTime = &t
+	}
+	if reason != "" {
+		patchData.Status.Reason = reason
+	}
+	if numCapturedPackets != 0 {
+		patchData.Status.NumCapturedPackets = numCapturedPackets
+	}
+	patchData.Status.PacketsFilePath = latestPC.Status.PacketsFilePath
```

Comment:
I think this should be reset to nil if the status is failed.

Reply:
updated